### PR TITLE
sql/schemachanger: only allow certain schema changes during logical replication

### PR DESCRIPTION
**Pull request by rafiss**
See individual commits. They're separated so the refactoring and test additions are easier to review.

---

If a table is being used as a destination by a logical replication
stream, we will now check if a schema change operation is in an
allowlist before executing it.

As we hear more use cases and deem other schema changes safe, we will
add to the allowlist over time.

Epic CRDB-38974
Release note: None

### CODE REVIEW

**Review by fqazi — CHANGES_REQUESTED** [review]
Looks, good just some very minor stuff

> Reviewed 16 of 16 files at r1, 26 of 26 files at r2, 6 of 6 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-:-O6gkk-6-xwPhm0EGPu0:b-j51fe1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rafiss, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/logical_data_replication.go` line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6gkRLa8PazytQ5BeXb:-O6gkRLa8PazytQ5BeXc:b-a1k5vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/cmd/roachtest/tests/logical_data_replication.go#L270)):*
> ```Go
> 	// Run allowlisted schema changes on the replicated table on both the left
> 	// and right sides.
> 	setup.left.sysSQL.Exec(t, fmt.Sprintf("ALTER TABLE %s.%s ADD COLUMN new_col INT NULL", ldrWorkload.dbName, ldrWorkload.tableName))
> ```

Is it worth running the successful case arbitrarily during the workload?
___
*[`pkg/sql/catalog/descpb/structured.proto` line 1224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6giYF08g93P-DuXhrF:-O6giYF08g93P-DuXhrG:bds9jhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/sql/catalog/descpb/structured.proto#L1224)):*
> ```proto
>   // LDRJobIDs is a list of logical data replication job IDs which reference
>   // this table as a destination for the replication stream.
>   repeated int64 ldr_job_ids = 62 [(gogoproto.nullable) = false, (gogoproto.customname) = "LDRJobIDs"];
> ```

Nit: Lets make this type safe 
```
(gogoproto.nullable) = false,  
(gogoproto.customname) = "JobID",  
(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb.JobID"\];
```
___
*[`pkg/sql/sem/tree/schema_helpers.go` line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6gkaPqAm3IFhDv25Tl:-O6gkaPqAm3IFhDv25Tm:bv4bgnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/sql/sem/tree/schema_helpers.go#L43)):*
> ```Go
> 	switch s := n.(type) {
> 	case *CreateIndex:
> 		return true
> ```

Do we want to allow unique indexes? Also do we want to allow ALTER INDEX as well?

File: pkg/sql/sem/tree/schema_helpers.go [hunk 1]
```diff
@@ -33,3 +33,26 @@ func IsSetOrResetSchemaLocked(n Statement) bool {
...
+// allowed to occur while the table is being referenced by a logical data
+// replication job as a destination table.
+func IsAllowedLDRSchemaChange(n Statement) bool {
+	switch s := n.(type) {
+	case *CreateIndex:
+		return true
+	case *DropIndex:
+		return true
+	case *AlterTable:
+		// Allow adding a nullable column without a default value if that is the
```
Comment (msbutler) [inline pkg/sql/sem/tree/schema_helpers.go:47]: could we just merge CreateIndex and DropIndex on the white list? There may be some LDR specific code we need to add for adding nullable columns without default values.

File: pkg/cmd/roachtest/tests/logical_data_replication.go [hunk 2]
```diff
@@ -220,6 +234,80 @@ func TestLDRBasic(ctx context.Context, t test.Test, c cluster.Cluster, setup mul
...
+		getLogicalDataReplicationJobInfo, t.Status, false /* tolerateErrors */)
+	defer llv.maybeLogLatencyHist()
+
+	rlv := makeLatencyVerifier("ldr-right", 0, maxExpectedLatency, t.L(),
+		getLogicalDataReplicationJobInfo, t.Status, false /* tolerateErrors */)
+	defer rlv.maybeLogLatencyHist()
+
+	// Run allowlisted schema changes on the replicated table on both the left
+	// and right sides.
+	setup.left.sysSQL.Exec(t, fmt.Sprintf("ALTER TABLE %s.%s ADD COLUMN new_col INT NULL", ldrWorkload.dbName, ldrWorkload.tableName))
```
Comment (msbutler) [inline pkg/cmd/roachtest/tests/logical_data_replication.go:270]: i feel like we should test the allow list with unit tests first and merge a roachtest later? See `logical_replication_job_test.go` for examples on setting up a unit test.

**Review by msbutler — COMMENTED** [review]
Thanks for plowing through this!!

**Comment by msbutler** [comment]
@rafiss also a heads up that your previous patch is printing warnings in `./dev build`:
```
WARNING: field TableDescriptor.LdrJobIds is a repeated non-nullable native type, nullable=false has no effect
```

File: pkg/sql/sem/tree/schema_helpers.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/sem/tree/schema_helpers.go:47]: > could we just merge CreateIndex and DropIndex on the white list? There may be some LDR specific code we need to add for adding nullable columns without default values.

not sure i follow. how does handling nullable columns relate to merging create/drop index in the allowlist?

File: pkg/cmd/roachtest/tests/logical_data_replication.go (hunk 2 above)
Comment (rafiss) [reply pkg/cmd/roachtest/tests/logical_data_replication.go:270]: thanks, that's the pointer i needed! i can also add logic tests that are SQL only

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @msbutler, @renatolabs, and @srosenberg)

___
*[`pkg/sql/catalog/descpb/structured.proto` line 1224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6giYF08g93P-DuXhrF:-O6l7Aib0bcwv8l-6H-G:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/sql/catalog/descpb/structured.proto#L1224)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Lets make this type safe 
```
(gogoproto.nullable) = false,  
(gogoproto.customname) = "JobID",  
(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb.JobID"\];
```
</blockquote></details>

done
___
*[`pkg/sql/sem/tree/schema_helpers.go` line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6gkaPqAm3IFhDv25Tl:-O6l7BJd5L8TLEOZnUBY:bbqkls6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/sql/sem/tree/schema_helpers.go#L43)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we want to allow unique indexes? Also do we want to allow ALTER INDEX as well?
</blockquote></details>

i think unique indexes should not be allowed actually, since it can make inserts fail

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @msbutler, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/logical_data_replication.go` line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6gkRLa8PazytQ5BeXb:-O6l7XvX36jNED4HDYE_:b9fmed6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/cmd/roachtest/tests/logical_data_replication.go#L270)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Is it worth running the successful case arbitrarily during the workload?
</blockquote></details>

could you rephrase? i don't think i follow which case you are referring to.

File: pkg/sql/sem/tree/schema_helpers.go (hunk 1 above)
Comment (msbutler) [reply pkg/sql/sem/tree/schema_helpers.go:47]: sorry, what i mean to say is: lets only include AddIndex and DropIndex on the white list for now.

Comment (rafiss) [reply pkg/sql/sem/tree/schema_helpers.go:47]: sounds good!

**Review by fqazi — COMMENTED** [review]
> Reviewed 13 of 13 files at r6, 9 of 9 files at r7, 7 of 7 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130697)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, @renatolabs, and @srosenberg)

**Review by fqazi — APPROVED** [review]
Great work! :lgtm_strong: 

> Reviewed 9 of 9 files at r9, 13 of 13 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-:-O6lGfD61p6NKYGOKtzi:buqj0om)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @msbutler, @rafiss, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/logical_data_replication.go` line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/130697#-O6gkRLa8PazytQ5BeXb:-O6lGZeS0BdoweoeaEqc:bkg9m7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ace86cfd5b2a852192366630b5824b1010af1ab4/pkg/cmd/roachtest/tests/logical_data_replication.go#L270)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you rephrase? i don't think i follow which case you are referring to.
</blockquote></details>

I just meant it might be worth running allowed schema changes with the workload concurrently.

**Comment by msbutler** [comment]
@rafiss a heads up that i'm currently working through describing a subset of CREATE INDEX schema changes that we'll need to disallow (e.g. the kv write path currently does not support replicating partial indexes). I think it's fine for now to merge the white list you have and we'll fine tune it in the coming weeks. Lemme give you tests a quick read also.

File: pkg/sql/sem/tree/schema_helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			stmt:      "CREATE INDEX idx ON t (a)",
+			isAllowed: true,
+		},
+		{
+			stmt:      "DROP INDEX idx",
+			isAllowed: true,
+		},
+		{
+			stmt:      "ALTER TABLE t ADD COLUMN a INT",
+			isAllowed: true,
```
Comment (msbutler) [inline pkg/sql/sem/tree/schema_helpers_test.go:40]: i think this test needs to be updated in this commit, as add nullable column is not allowed yet

File: pkg/sql/sqlerrors/errors.go [hunk 4]
```diff
@@ -52,6 +53,24 @@ func NewSchemaChangeOnLockedTableErr(tableName string) error {
...
+	for i, v := range jobIDs {
+		ids[i] = strconv.Itoa(int(v))
+	}
+	return errors.WithHintf(
+		pgerror.Newf(
+			pgcode.FeatureNotSupported,
+			"this schema change is disallowed on table %s because it is referenced by "+
+				"one or more logical replication jobs [%s]", tableName, strings.Join(ids, ", "),
+		),
+		"To perform the schema change, cancel the logical replication job, then execute "+
```
Comment (msbutler) [inline pkg/sql/sqlerrors/errors.go:71]: hrm, for certain schema changes, this is the right advice, but for others, the sc may make LDR impossible resume. How about we remove this directive for now, and rely on docs for how a user should run an unsupported SC?

**Review by msbutler — APPROVED** [review]
LGTM modulo the test failures are trivial to fix! Looks like for `TestForeignKeyConstraints`, you just need to cancel the LDR job if it is successfully created.

File: pkg/sql/sem/tree/schema_helpers_test.go (hunk 3 above)
Comment (rafiss) [reply pkg/sql/sem/tree/schema_helpers_test.go:40]: done

File: pkg/sql/sqlerrors/errors.go (hunk 4 above)
Comment (rafiss) [reply pkg/sql/sqlerrors/errors.go:71]: sounds good!

**Comment by rafiss** [comment]
tftr!

bors r+
